mozilla/kuma
# Bug 1150118 - lazyload fonts

First pass at lazy loading fonts.  Only lazy loading `Open Sans` and `Open Sans Light` at the moment.

How it works:
1. User visits for first time
2. User does not have right data attributes on `html` element so webfonts do not display
3. Javascript runs loading as many webfonts as possible before either a) downloading all and displaying them, or  b) timing out at 3 seconds and displaying what it has loaded so far. When all fonts are loaded a cookie is set.
4. User goes to new page, if the cookie is set the server includes the data attributes on the `html` tag and webfonts display immediately (if cookie is not set javascript runs again, attempting to load all fonts again)

To test:
- clear cache
- load page, watch for: fallback displays immediately, site and heading fonts swap in simultaneously
- reload page without clearing cache, watch for: site and heading fonts should display immediately
- get a really slow connection
- clear cache
- load page, watch for: on a slow enough connection the fonts should time out, the page should never swap in the site and heading fonts
- reload page without clearing cache, watch for: site and heading fonts may display immediately or may time out again (watch the network panel, if they loaded they should display)

I'm really happy with this. I think it is an improvement on the current first load user experience :D 

Current (prod) in Firefox:
![prod](https://cloud.githubusercontent.com/assets/854701/7243586/a20be926-e783-11e4-89bc-7421bd7e2294.gif)

Lazyload in Firefox:
![ffo](https://cloud.githubusercontent.com/assets/854701/7243587/aa96a0cc-e783-11e4-9086-6b806d0b9ba7.gif)

I know everyone is really skeptical that we should be messing with the experience so I wanted to get some feedback before solving the edge cases.

To do:
- [x] waffle
- [x] check license on FontFaceObserver is compatible with MDN
- [x] find solution to revert to default behavior if JavaScript disabled
- [x] see if anything can be done about hard refresh load when cookie set
- [x] check all uses of `Open Sans` and `Open Sans Light` have fallbacks declared.

To do that might have to wait for a waffled launch:
- [ ] get feedback on user experience to decide on go-ahead
- [ ] test in IE
